kornia/feature/keynet.py

```diff
@@ -0,0 +1,315 @@
+        self.eval()
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        x - input image
+        """
+        shape_im = x.shape
+        feats: List[torch.Tensor] = [self.feature_extractor(x)]
+        for i in range(1, self.num_levels):
+            x = pyrdown(x, factor=1.2)
+            feats_i = self.feature_extractor(x)
+            feats_i = F.interpolate(feats_i, size=(shape_im[2], shape_im[3]), mode='bilinear')
+            feats.append(feats_i)
+        scores = self.last_conv(torch.cat(feats, dim=1))
+        return scores
+
+
+class KeyNetDetector(nn.Module):
+    ''' Multi-scale feature detector based on KeyNet. This is based on the original code from paper
+    "Key.Net: Keypoint Detection by Handcrafted and Learned CNN Filters".
+    See :cite:`KeyNet2019` for more details.
+
+    Args:
+        pretrained: Download and set pretrained weights to the model.
+        num_features: Number of features to detect.
+        keynet_conf: Dict with initiliazation parameters. Do not pass it, unless you know what you are doing`.
+        ori_module: for local feature orientation estimation. Default:class:`~kornia.feature.PassLAF`,
+           which does nothing. See :class:`~kornia.feature.LAFOrienter` for details.
+        aff_module: for local feature affine shape estimation. Default: :class:`~kornia.feature.PassLAF`,
+            which does nothing. See :class:`~kornia.feature.LAFAffineShapeEstimator` for details.
+    '''
+    def __init__(self,
+                 pretrained: bool = False,
+                 num_features: int = 2048,
+                 keynet_conf: Dict = keynet_config['KeyNet_default_config'],
+                 ori_module: nn.Module = PassLAF(),
+                 aff_module: nn.Module = PassLAF()):
+        super().__init__()
+        self.model = KeyNet(pretrained, keynet_conf)
+        # Load extraction configuration
+        self.num_pyramid_levels: int = keynet_conf['pyramid_levels']
+        self.num_upscale_levels: int = keynet_conf['up_levels']
+        self.scale_factor_levels: float = keynet_conf['scale_factor_levels']
+        self.mr_size: float = keynet_conf['s_mult']
+        self.nms_size: int = keynet_conf['nms_size']
+        self.nms = NonMaximaSuppression2d((self.nms_size, self.nms_size))
+        self.num_features = num_features
+        self.ori = ori_module
+        self.aff = aff_module
+
+    def remove_borders(self, score_map, borders: int = 15):
+        """It removes the borders of the image to avoid detections on the corners."""
+        shape = score_map.shape
+        mask = torch.ones_like(score_map)
+
+        mask[:, :, 0:borders, :] = 0
+        mask[:, :, :, 0:borders] = 0
+        mask[:, :, shape[2] - borders:shape[2], :] = 0
+        mask[:, :, :, shape[3] - borders:shape[3]] = 0
+        return mask * score_map
+
+    def detect_features_on_single_level(self, level_img: torch.Tensor, num_kp: int, factor: Tuple[float, float]):
```

Comment:
```suggestion
    def detect_features_on_single_level(self, level_img: torch.Tensor, num_kp: int, factor: Tuple[float, float]):
```
```suggestion
    def detect_features_on_single_level(self, level_img: torch.Tensor, num_kp: int, factor: Tuple[float, float]) -> Tuple[Tensor, Tensor]:
```
NOTE: i'm starting to `from torch import Tensor` to make easier later to switch torch/kornia